File: src/lib/engine.ts
export type GameId = 'super_lotto_638' | 'lotto_649' | 'daily_cash_539';

export interface GameSchema {
  id: GameId;
  pool: number;
  count: number;
  hasSpecial: boolean;
  specialPool: number;
}

export const GAME_SCHEMAS: Record<GameId, GameSchema> = {
  super_lotto_638: { id: 'super_lotto_638', pool: 38, count: 6, hasSpecial: true, specialPool: 8 },
  lotto_649: { id: 'lotto_649', pool: 49, count: 6, hasSpecial: true, specialPool: 49 }, // actually from 43 remaining balls, but numbered 1-49
  daily_cash_539: { id: 'daily_cash_539', pool: 39, count: 5, hasSpecial: false, specialPool: 0 }
};

export interface DrawRecord {
  draw_id: string;
  game_type: string;
  draw_date: string;
  numbers: number[];
  special_number: number | null;
}

export interface PredictionSet {
  numbers: number[];
  special: number | null;
  justification: string;
  riskProfile: string;
  narrative?: string; // Filled by Gemini
}

export interface EngineResult {
  alpha: PredictionSet;
  beta: PredictionSet;
  gamma: PredictionSet;
  metrics: {
    targetSum: number;
    hotCount: number;
    coldCount: number;
    repeatProbability: number;
  };
}

// ---------------------------------------------------------
// Math Utilities
// ---------------------------------------------------------

export function mulberry32(a: number) {
  return function() {
    let t = a += 0x6D2B79F5;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }
}

export function hashString(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash;
  }
  return hash >>> 0;
}

const sumArray = (arr: number[]) => arr.reduce((a, b) => a + b, 0);

export function pickRandom(arr: number[], n: number, rng: () => number) {
  const shuffled = [...arr];
  for (let i = shuffled.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [shuffled[i], shuffled[j]] = [shuffled[j] as number, shuffled[i] as number];
  }
  return shuffled.slice(0, n).sort((a, b) => a - b);
}

// ---------------------------------------------------------
// Core Engine Analysis
// ---------------------------------------------------------

export function runPredictionEngine(gameId: GameId, rawDraws: DrawRecord[]): EngineResult {
  const schema = GAME_SCHEMAS[gameId];
  if (!schema) throw new Error("Invalid Game ID");

  // Filter valid draws and guarantee exactly 50 max
  const sortedDraws = [...rawDraws]
    .sort((a, b) => new Date(b.draw_date).getTime() - new Date(a.draw_date).getTime())
    .filter(d => d.numbers && d.numbers.length === schema.count)
    .slice(0, 50);

  if (sortedDraws.length === 0) {
    throw new Error("No valid draws found for " + gameId);
  }

  // Seed deterministic RNG
  const latestDraw = sortedDraws[0]!;
  const seedString = `${gameId}_${latestDraw.draw_id}_${latestDraw.draw_date}`;
  const rng = mulberry32(hashString(seedString));

  // --- 1. Frequency, Recency, and Gap Analysis ---
  const stats = new Map<number, { count: number, recencyWeight: number, gap: number }>();
  for (let i = 1; i <= schema.pool; i++) {
    stats.set(i, { count: 0, recencyWeight: 0, gap: -1 });
  }

  const reversedDraws = [...sortedDraws].reverse();
  reversedDraws.forEach((draw, index) => {
    const decay = Math.pow(0.5, (reversedDraws.length - 1 - index) / 15); // half-life 15 draws
    stats.forEach(val => val.gap++);
    draw.numbers.forEach(num => {
      const s = stats.get(num)!;
      s.count++;
      s.recencyWeight += decay;
      s.gap = 0;
    });
  });

  const sortedByScore = Array.from(stats.entries()).sort((a, b) => b[1].recencyWeight - a[1].recencyWeight);
  const hotNumbers = sortedByScore.slice(0, Math.ceil(schema.pool * 0.2)).map(e => e[0]);
  
  const sortedByGap = Array.from(stats.entries()).sort((a, b) => b[1].gap - a[1].gap);
  const coldNumbers = sortedByGap.slice(0, Math.ceil(schema.pool * 0.2)).map(e => e[0]); 

  // --- 2. Pair Correlation ---
  const pairCounts = new Map<string, number>();
  sortedDraws.forEach(draw => {
    const nums = draw.numbers.sort((a, b) => a - b);
    for (let i = 0; i < nums.length; i++) {
      for (let j = i + 1; j < nums.length; j++) {
        const key = `${nums[i]},${nums[j]}`;
        pairCounts.set(key, (pairCounts.get(key) || 0) + 1);
      }
    }
  });

  const strongPairs = Array.from(pairCounts.entries())
    .filter(e => e[1] > 2)
    .sort((a, b) => b[1] - a[1]);

  // --- 3. Poisson Repeat Analysis ---
  let repeatsFromPrev = 0;
  for (let i = 0; i < sortedDraws.length - 1; i++) {
    const curr = sortedDraws[i]!.numbers;
    const prev = sortedDraws[i+1]!.numbers;
    const common = curr.filter(x => prev.includes(x)).length;
    if (common > 0) repeatsFromPrev++;
  }
  const empiricalRepeatProb = repeatsFromPrev / Math.max(1, sortedDraws.length - 1);
  const lambda = Math.pow(schema.count, 2) / schema.pool;
  const theoreticalRepeatProb = 1 - Math.exp(-lambda);
  const repeatProbability = ((empiricalRepeatProb + theoreticalRepeatProb) / 2) * 100;

  // --- 4. Monte Carlo Simulation for "Balanced Set" ---
  const meanSum = sumArray(sortedDraws.map(d => sumArray(d.numbers))) / sortedDraws.length;
  const meanOdd = sumArray(sortedDraws.map(d => d.numbers.filter(n => n % 2 !== 0).length)) / sortedDraws.length;
  const meanHigh = sumArray(sortedDraws.map(d => d.numbers.filter(n => n > schema.pool / 2).length)) / sortedDraws.length;

  let bestAlpha: number[] = [];
  let bestAlphaError = Infinity;
  const fullPool = Array.from({ length: schema.pool }, (_, i) => i + 1);
  
  for (let i = 0; i < 100000; i++) {
    const candidate = pickRandom(fullPool, schema.count, rng);
    
    // Entropy guard
    let consecutives = 0;
    for(let j=0; j<candidate.length-1; j++) {
      if(candidate[j+1]! - candidate[j]! === 1) consecutives++;
    }
    if (consecutives > 2) continue;

    const sum = sumArray(candidate);
    const odds = candidate.filter(n => n % 2 !== 0).length;
    const highs = candidate.filter(n => n > schema.pool / 2).length;

    const error = Math.abs(sum - meanSum) / meanSum + 
                  Math.abs(odds - meanOdd) / meanOdd + 
                  Math.abs(highs - meanHigh) / meanHigh;
    
    if (error < bestAlphaError) {
      bestAlphaError = error;
      bestAlpha = candidate;
    }
  }

  // --- 5. Generate Sets ---

  const getSpecial = (rngFunc: () => number) => {
    if (!schema.hasSpecial) return null;
    return Math.floor(rngFunc() * schema.specialPool) + 1;
  };

  const alpha = {
    numbers: bestAlpha,
    special: getSpecial(rng),
    justification: `Monte Carlo optimal set. Sum: ${sumArray(bestAlpha)} (Target: ${Math.round(meanSum)}). Matches historical 50-draw means for Odd/Even and High/Low splits while preserving spatial entropy.`,
    riskProfile: 'Low Variance - Converges to Mean',
  };

  // Beta (Momentum)
  let betaNums = new Set<number>();
  if (repeatProbability > 50 && sortedDraws[0]!.numbers.length > 0) {
    const lastDrawNums = sortedDraws[0]!.numbers;
    betaNums.add(lastDrawNums[Math.floor(rng() * lastDrawNums.length)]!);
  }
  
  if (strongPairs.length > 0) {
    const topPair = strongPairs[0]![0]!.split(',').map(Number);
    betaNums.add(topPair[0]!);
    betaNums.add(topPair[1]!);
  }

  let hotIdx = 0;
  while(betaNums.size < schema.count && hotIdx < hotNumbers.length) {
    betaNums.add(hotNumbers[hotIdx]!);
    hotIdx++;
  }
  while(betaNums.size < schema.count) {
    betaNums.add(Math.floor(rng() * schema.pool) + 1);
  }

  const betaArray = Array.from(betaNums).sort((a,b)=>a-b);
  const beta = {
    numbers: betaArray,
    special: getSpecial(rng),
    justification: `Momentum selection based on Top 20% exponentially decayed frequency, integrated with a ${repeatProbability.toFixed(1)}% Poisson repeat expectation and historical pair correlations.`,
    riskProfile: 'High Momentum - Trend Following',
  };

  // Gamma (Chaos)
  const isOddChaos = rng() > 0.5;
  const chaosPool = coldNumbers.filter(n => (n % 2 !== 0) === isOddChaos);
  let gammaArray = pickRandom(chaosPool.length >= schema.count ? chaosPool : coldNumbers, schema.count, rng);
  
  if (gammaArray.length < schema.count) {
    gammaArray = pickRandom(fullPool, schema.count, rng);
  }

  const gamma = {
    numbers: gammaArray,
    special: getSpecial(rng),
    justification: `Black Swan pattern break. Built primarily from high-omission (Cold) numbers structured as an extreme ${isOddChaos ? 'All-Odd' : 'All-Even'} topological split.`,
    riskProfile: 'Extreme - Pattern Breaking',
  };

  return {
    alpha,
    beta,
    gamma,
    metrics: {
      targetSum: Math.round(meanSum),
      hotCount: hotNumbers.length,
      coldCount: coldNumbers.length,
      repeatProbability
    }
  };
}

File: numberd_app/lib/core/utils/engine.dart
import 'dart:math' as math;
import '../models/draw_record.dart';

class GameSchema {
  final String id;
  final int pool;
  final int count;
  final bool hasSpecial;
  final int specialPool;

  const GameSchema({
    required this.id,
    required this.pool,
    required this.count,
    required this.hasSpecial,
    required this.specialPool,
  });
}

const gameSchemas = {
  'super_lotto_638': GameSchema(id: 'super_lotto_638', pool: 38, count: 6, hasSpecial: true, specialPool: 8),
  'lotto_649': GameSchema(id: 'lotto_649', pool: 49, count: 6, hasSpecial: true, specialPool: 49),
  'daily_cash_539': GameSchema(id: 'daily_cash_539', pool: 39, count: 5, hasSpecial: false, specialPool: 0),
};

class PredictionSet {
  final List<int> numbers;
  final int? special;
  final String justification;
  final String riskProfile;

  PredictionSet({
    required this.numbers,
    this.special,
    required this.justification,
    required this.riskProfile,
  });
}

class EngineResult {
  final PredictionSet alpha;
  final PredictionSet beta;
  final PredictionSet gamma;
  final int targetSum;
  final int hotCount;
  final int coldCount;
  final double repeatProbability;

  EngineResult({
    required this.alpha,
    required this.beta,
    required this.gamma,
    required this.targetSum,
    required this.hotCount,
    required this.coldCount,
    required this.repeatProbability,
  });
}

int imul(int a, int b) {
  int aHi = (a >>> 16) & 0xffff;
  int aLo = a & 0xffff;
  int bHi = (b >>> 16) & 0xffff;
  int bLo = b & 0xffff;
  return ((aLo * bLo) + (((aHi * bLo + aLo * bHi) << 16) >>> 0)) & 0xffffffff;
}

class PRNG {
  int _state;

  PRNG(this._state);

  double nextDouble() {
    _state = (_state + 0x6D2B79F5) & 0xFFFFFFFF;
    int t = _state;
    t = imul(t ^ (t >>> 15), t | 1);
    t ^= t + imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296.0;
  }
}

int hashString(String str) {
  int hash = 0;
  for (int i = 0; i < str.length; i++) {
    final char = str.codeUnitAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash;
  }
  return hash >>> 0;
}

List<int> pickRandom(List<int> arr, int n, PRNG rng) {
  if (arr.isEmpty) return [];
  if (arr.length <= n) {
    final res = List<int>.from(arr);
    res.sort();
    return res;
  }
  final shuffled = List<int>.from(arr);
  for (int i = shuffled.length - 1; i > 0; i--) {
    final j = (rng.nextDouble() * (i + 1)).floor();
    final temp = shuffled[i];
    shuffled[i] = shuffled[j];
    shuffled[j] = temp;
  }
  final res = shuffled.sublist(0, n);
  res.sort();
  return res;
}

int sumArray(List<int> arr) => arr.fold(0, (a, b) => a + b);

EngineResult runPredictionEngine(String gameId, List<DrawRecord> rawDraws) {
  final schema = gameSchemas[gameId];
  if (schema == null) throw ArgumentError("Invalid Game ID");

  final sortedDraws = List<DrawRecord>.from(rawDraws)
    ..sort((a, b) => DateTime.parse(b.drawDate).compareTo(DateTime.parse(a.drawDate)));
  
  // Filter valid draws: ensure exact count and that all numbers are within pool boundaries
  final validDraws = sortedDraws.where((d) => 
    d.numbers.length == schema.count &&
    d.numbers.every((n) => n >= 1 && n <= schema.pool)
  ).take(50).toList();

  if (validDraws.isEmpty) {
    throw StateError("No valid draws found for $gameId");
  }

  final latestDraw = validDraws.first;
  final seedString = "${gameId}_${latestDraw.drawId}_${latestDraw.drawDate}";
  final rng = PRNG(hashString(seedString));

  // --- 1. Frequency, Recency, and Gap Analysis ---
  final stats = <int, _Stat>{};
  for (int i = 1; i <= schema.pool; i++) {
    stats[i] = _Stat(count: 0, recencyWeight: 0, gap: -1);
  }

  final reversedDraws = validDraws.reversed.toList();
  for (int index = 0; index < reversedDraws.length; index++) {
    final draw = reversedDraws[index];
    final decay = math.pow(0.5, (reversedDraws.length - 1 - index) / 15.0).toDouble();
    
    for (final s in stats.values) {
      s.gap++;
    }
    
    for (final num in draw.numbers) {
      final s = stats[num];
      if (s != null) {
        s.count++;
        s.recencyWeight += decay;
        s.gap = 0;
      }
    }
  }

  final sortedByScore = stats.entries.toList()
    ..sort((a, b) => b.value.recencyWeight.compareTo(a.value.recencyWeight));
  final hotCount = (schema.pool * 0.2).ceil();
  final hotNumbers = sortedByScore.take(hotCount).map((e) => e.key).toList();

  final sortedByGap = stats.entries.toList()
    ..sort((a, b) => b.value.gap.compareTo(a.value.gap));
  final coldCount = (schema.pool * 0.2).ceil();
  final coldNumbers = sortedByGap.take(coldCount).map((e) => e.key).toList();

  // --- 2. Pair Correlation ---
  final pairCounts = <String, int>{};
  for (final draw in validDraws) {
    final nums = List<int>.from(draw.numbers)..sort();
    for (int i = 0; i < nums.length; i++) {
      for (int j = i + 1; j < nums.length; j++) {
        final key = '${nums[i]},${nums[j]}';
        pairCounts[key] = (pairCounts[key] ?? 0) + 1;
      }
    }
  }

  final strongPairs = pairCounts.entries.where((e) => e.value > 2).toList()
    ..sort((a, b) => b.value.compareTo(a.value));

  // --- 3. Poisson Repeat Analysis ---
  int repeatsFromPrev = 0;
  for (int i = 0; i < validDraws.length - 1; i++) {
    final curr = validDraws[i].numbers;
    final prev = validDraws[i + 1].numbers;
    final common = curr.where((x) => prev.contains(x)).length;
    if (common > 0) repeatsFromPrev++;
  }
  final empiricalRepeatProb = repeatsFromPrev / math.max(1, validDraws.length - 1);
  final lambda = math.pow(schema.count, 2) / schema.pool;
  final theoreticalRepeatProb = 1 - math.exp(-lambda);
  final repeatProbability = ((empiricalRepeatProb + theoreticalRepeatProb) / 2) * 100;

  // --- 4. Monte Carlo Simulation for "Balanced Set" ---
  final meanSum = validDraws.map((d) => sumArray(d.numbers)).reduce((a, b) => a + b) / validDraws.length;
  final meanOdd = validDraws.map((d) => d.numbers.where((n) => n % 2 != 0).length).reduce((a, b) => a + b) / validDraws.length;
  final meanHigh = validDraws.map((d) => d.numbers.where((n) => n > schema.pool / 2).length).reduce((a, b) => a + b) / validDraws.length;

  List<int> bestAlpha = [];
  double bestAlphaError = double.infinity;
  final fullPool = List.generate(schema.pool, (i) => i + 1);

  for (int i = 0; i < 100000; i++) {
    final candidate = pickRandom(fullPool, schema.count, rng);
    
    int consecutives = 0;
    for (int j = 0; j < candidate.length - 1; j++) {
      if (candidate[j + 1] - candidate[j] == 1) consecutives++;
    }
    if (consecutives > 2) continue;

    final sum = sumArray(candidate);
    final odds = candidate.where((n) => n % 2 != 0).length;
    final highs = candidate.where((n) => n > schema.pool / 2).length;

    final error = (sum - meanSum).abs() / meanSum + 
                  (odds - meanOdd).abs() / meanOdd + 
                  (highs - meanHigh).abs() / meanHigh;
    
    if (error < bestAlphaError) {
      bestAlphaError = error;
      bestAlpha = candidate;
    }
  }

  // --- 5. Generate Sets ---
  int? getSpecial(PRNG rngFunc) {
    if (!schema.hasSpecial || schema.specialPool <= 0) return null;
    return (rngFunc.nextDouble() * schema.specialPool).floor() + 1;
  }

  final alpha = PredictionSet(
    numbers: bestAlpha,
    special: getSpecial(rng),
    justification: 'Monte Carlo optimal set. Sum: ${sumArray(bestAlpha)} (Target: ${meanSum.round()}). Matches historical 50-draw means for Odd/Even and High/Low splits while preserving spatial entropy.',
    riskProfile: 'Low Variance - Converges to Mean',
  );

  // Beta (Momentum)
  final betaNums = <int>{};
  if (repeatProbability > 50 && validDraws[0].numbers.isNotEmpty) {
    final lastDrawNums = validDraws[0].numbers;
    betaNums.add(lastDrawNums[(rng.nextDouble() * lastDrawNums.length).floor()]);
  }
  
  if (strongPairs.isNotEmpty) {
    final topPair = strongPairs[0].key.split(',').map(int.parse).toList();
    if (topPair[0] >= 1 && topPair[0] <= schema.pool) betaNums.add(topPair[0]);
    if (topPair[1] >= 1 && topPair[1] <= schema.pool) betaNums.add(topPair[1]);
  }

  int hotIdx = 0;
  while (betaNums.length < schema.count && hotIdx < hotNumbers.length) {
    betaNums.add(hotNumbers[hotIdx]);
    hotIdx++;
  }
  while (betaNums.length < schema.count) {
    betaNums.add((rng.nextDouble() * schema.pool).floor() + 1);
  }

  final betaArray = betaNums.toList()..sort();
  final beta = PredictionSet(
    numbers: betaArray,
    special: getSpecial(rng),
    justification: 'Momentum selection based on Top 20% exponentially decayed frequency, integrated with a ${repeatProbability.toStringAsFixed(1)}% Poisson repeat expectation and historical pair correlations.',
    riskProfile: 'High Momentum - Trend Following',
  );

  // Gamma (Chaos)
  final isOddChaos = rng.nextDouble() > 0.5;
  final chaosPool = coldNumbers.where((n) => (n % 2 != 0) == isOddChaos).toList();
  List<int> gammaArray = pickRandom(chaosPool.length >= schema.count ? chaosPool : coldNumbers, schema.count, rng);
  
  if (gammaArray.length < schema.count) {
    gammaArray = pickRandom(fullPool, schema.count, rng);
  }

  final gamma = PredictionSet(
    numbers: gammaArray,
    special: getSpecial(rng),
    justification: 'Black Swan pattern break. Built primarily from high-omission (Cold) numbers structured as an extreme ${isOddChaos ? 'All-Odd' : 'All-Even'} topological split.',
    riskProfile: 'Extreme - Pattern Breaking',
  );

  return EngineResult(
    alpha: alpha,
    beta: beta,
    gamma: gamma,
    targetSum: meanSum.round(),
    hotCount: hotNumbers.length,
    coldCount: coldNumbers.length,
    repeatProbability: repeatProbability,
  );
}

class _Stat {
  int count;
  double recencyWeight;
  int gap;

  _Stat({required this.count, required this.recencyWeight, required this.gap});
}

